### PR TITLE
ET factories for stablecoins

### PR DESCRIPTION
**Pull request by rkolpakov**

### CODE REVIEW

File: contracts/AllowedRecipientsBuilder.sol [hunk 1]
```diff
@@ -155,185 +175,159 @@ contract AllowedRecipientsBuilder {
...
-        assert(registry.hasRole(SET_PARAMETERS_ROLE, admin));
-        assert(registry.hasRole(UPDATE_SPENT_AMOUNT_ROLE, admin));
-        assert(registry.hasRole(DEFAULT_ADMIN_ROLE, admin));
+        assert(recipientRegistry.hasRole(ADD_RECIPIENT_TO_ALLOWED_LIST_ROLE, admin));
+        assert(recipientRegistry.hasRole(REMOVE_RECIPIENT_FROM_ALLOWED_LIST_ROLE, admin));
+        assert(recipientRegistry.hasRole(SET_PARAMETERS_ROLE, admin));
+        assert(recipientRegistry.hasRole(UPDATE_SPENT_AMOUNT_ROLE, admin));
+        assert(recipientRegistry.hasRole(DEFAULT_ADMIN_ROLE, admin));
+        assert(tokenRegistry.hasRole(ADD_TOKEN_TO_ALLOWED_LIST_ROLE, admin));
+        assert(tokenRegistry.hasRole(REMOVE_TOKEN_FROM_ALLOWED_LIST_ROLE, admin));
```
Comment (Psirex) [inline contracts/AllowedRecipientsBuilder.sol:224]: Seems like the check for DEFAULT_ADMIN_ROLE for the tokensRegistry is missing here and below

File: contracts/AllowedRecipientsBuilder.sol [hunk 2]
```diff
@@ -155,185 +175,159 @@ contract AllowedRecipientsBuilder {
...
 
         assert(topUpAllowedRecipients.token() == _token);
         assert(topUpAllowedRecipients.finance() == finance);
         assert(topUpAllowedRecipients.easyTrack() == easyTrack);
         assert(topUpAllowedRecipients.trustedCaller() == _trustedCaller);
-        assert(
-            address(topUpAllowedRecipients.allowedRecipientsRegistry()) ==
-                _allowedRecipientsRegistry
-        );
+        assert(address(topUpAllowedRecipients.allowedRecipientsRegistry()) == _allowedRecipientsRegistry);
```
Comment (Psirex) [inline contracts/AllowedRecipientsBuilder.sol:264]: Also, need the check for the _allowedTokensRegistry

File: tests/evm_script_factories/test_top_up_allowed_recipients.py [hunk 3]
```diff
@@ -5,47 +5,62 @@
...
+        recipients_registry,
+        tokens_registry,
+        finance,
+        ldo,
+        easy_track,
     )
 
     assert top_up_factory.token() == ldo
-    assert top_up_factory.allowedRecipientsRegistry() == registry
+    assert top_up_factory.allowedRecipientsRegistry() == recipients_registry
```
Comment (Psirex) [inline tests/evm_script_factories/test_top_up_allowed_recipients.py:38]: Also need a check for the tokens registry

File: tests/integration/test_allowed_recipients_happy_path.py [hunk 4]
```diff
@@ -92,12 +109,14 @@ def full_setup(
...
     deployer,
+    ldo,
+    usdc
 ):
     deploy_tx = allowed_recipients_builder.deployFullSetup(
         trusted_caller,
-        lido_contracts.ldo,
         allowed_recipients_default_params.limit,
         allowed_recipients_default_params.period_duration_months,
+        [ldo, usdc],
```
Comment (Psirex) [inline tests/integration/test_allowed_recipients_happy_path.py:119]: It's better to use tokens we actually will use together in the same tokens registry

File: tests/evm_script_factories/test_top_up_allowed_recipients.py [hunk 5]
```diff
@@ -162,20 +196,65 @@ def test_fail_create_evm_script_if_recipient_not_allowed(
...
+    tokens_registry.addToken(top_up_factory.token(), {"from": add_token_role_holder})
+
     with reverts("RECIPIENT_NOT_ALLOWED"):
-        top_up_factory.createEVMScript(trusted_caller, make_call_data([stranger.address], [123]))
+        top_up_factory.createEVMScript(
+            trusted_caller, make_call_data([stranger.address], [123])
+        )
+
+
+def test_fail_create_evm_script_if_token_not_allowed(
```
Comment (Psirex) [inline tests/evm_script_factories/test_top_up_allowed_recipients.py:218]: Would be nice to have a test for multiple tokens in the registry

File: tests/test_allowed_tokens_registry.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+def test_normalize_amount(allowed_tokens_registry):
+    (registry, _, _, _) = allowed_tokens_registry
+
+    erc20decimals18 = MockERC20.deploy(18, {"from": accounts[0]})
+
+    with reverts("TOKEN_ADDRESS_IS_ZERO"):
+        registry.normalizeAmount(1, ZERO_ADDRESS)
+
+    
+    amount1 = 1000000000000000000
```
Comment (Psirex) [inline tests/test_allowed_tokens_registry.py:234]: Better to use non-round numbers, to check that rounding works correctly.

File: tests/evm_script_factories/test_top_up_allowed_recipients.py [hunk 7]
```diff
@@ -162,20 +192,65 @@ def test_fail_create_evm_script_if_recipient_not_allowed(
...
     registry.setLimitParameters(int(100e18), 12, {"from": set_limit_role_holder})
 
     top_up_factory = owner.deploy(
-        TopUpAllowedRecipients, trusted_caller, registry, finance, ldo, easy_track
+        TopUpAllowedRecipients, trusted_caller, registry, tokens_registry, finance, easy_track
     )
 
+    tokens_registry.addToken(ldo, {"from": add_token_role_holder})
+
     with reverts("RECIPIENT_NOT_ALLOWED"):
```
Comment (Psirex) [inline tests/evm_script_factories/test_top_up_allowed_recipients.py:208]: Also we need test for zero address recipient

File: tests/evm_script_factories/test_top_up_allowed_recipients.py [hunk 8]
```diff
@@ -263,27 +352,35 @@ def test_fail_create_evm_script_if_sum_exceeds_limit(
...
+        TopUpAllowedRecipients, owner, registry, tokens_registry, finance, easy_track
     )
 
+    tokens_registry.addToken(ldo, {"from": add_token_role_holder})
+
     with reverts("SUM_EXCEEDS_SPENDABLE_BALANCE"):
         top_up_factory.createEVMScript(owner, call_data)
 
 
 def test_create_evm_script_correctly(
```
Comment (Psirex) [inline tests/evm_script_factories/test_top_up_allowed_recipients.py:375]: It's better to use a token with decimals that differ from the recipients registry. To validate that normalization also happens

File: tests/test_allowed_recipients_builder.py [hunk 9]
```diff
@@ -182,62 +181,76 @@ def test_deploy_allowed_recipients_registry(
...
     titles = ["account 3", "account 4"]
     spentAmount = 1e10
 
-    tx = allowed_recipients_builder.deployAllowedRecipientsRegistry(
+    tx_recipients = allowed_recipients_builder.deployAllowedRecipientsRegistry(
         limit, period, recipients, titles, spentAmount, True, {"from": stranger}
     )
 
-    registry_address = tx.events["AllowedRecipientsRegistryDeployed"][
+    tx_tokens = allowed_recipients_builder.deployAllowedTokensRegistry(tokens, {"from": stranger})
```
Comment (Psirex) [inline tests/test_allowed_recipients_builder.py:200]: Move checks for allowed tokens registry in a standalone test. Similar to "test_deploy_allowed_recipients_registry", check that:
- roles were granted correctly
- builder doesn't have any roles granted
- all tokens were added

File: tests/test_allowed_recipients_builder.py [hunk 10]
```diff
@@ -339,46 +361,50 @@ def test_deploy_full_setup(
...
+        REMOVE_RECIPIENT_FROM_ALLOWED_LIST_ROLE, recipients_registry_address
     )
-    assert not registry.hasRole(SET_PARAMETERS_ROLE, registry_address)
-    assert not registry.hasRole(UPDATE_SPENT_AMOUNT_ROLE, registry_address)
-    assert not registry.hasRole(DEFAULT_ADMIN_ROLE, registry_address)
+    assert not recipients_registry.hasRole(SET_PARAMETERS_ROLE, recipients_registry_address)
+    assert not recipients_registry.hasRole(UPDATE_SPENT_AMOUNT_ROLE, recipients_registry_address)
+    assert not recipients_registry.hasRole(DEFAULT_ADMIN_ROLE, recipients_registry_address)
+
+    assert tokens_registry.hasRole(ADD_TOKEN_TO_ALLOWED_LIST_ROLE, agent)
```
Comment (Psirex) [inline tests/test_allowed_recipients_builder.py:403]: Missing check that tokens were added

File: tests/test_allowed_recipients_builder.py [hunk 11]
```diff
@@ -449,3 +481,9 @@ def test_deploy_deploy_single_recipient_top_up_only_setup(
     assert not registry.hasRole(SET_PARAMETERS_ROLE, registry_address)
     assert not registry.hasRole(UPDATE_SPENT_AMOUNT_ROLE, registry_address)
     assert not registry.hasRole(DEFAULT_ADMIN_ROLE, registry_address)
+
+    assert tokens_registry.hasRole(ADD_TOKEN_TO_ALLOWED_LIST_ROLE, agent)
```
Comment (Psirex) [inline tests/test_allowed_recipients_builder.py:485]: Check for added tokens

File: tests/test_allowed_recipients_builder.py (hunk 10 above)
Comment (Psirex) [reply tests/test_allowed_recipients_builder.py:403]: Also missing the check for DEFAULT_ADMIN_ROLE

File: tests/test_allowed_tokens_registry.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+        [remove_token_role_holder],
+    )
+
+    assert registry.hasRole(ADD_TOKEN_TO_ALLOWED_LIST_ROLE, add_token_role_holder)
+    assert registry.hasRole(
+        REMOVE_TOKEN_FROM_ALLOWED_LIST_ROLE, remove_token_role_holder
+    )
+
+    for role_holder in [add_token_role_holder, remove_token_role_holder]:
+        assert not registry.hasRole(registry.DEFAULT_ADMIN_ROLE(), role_holder)
```
Comment (Psirex) [inline tests/test_allowed_tokens_registry.py:33]: Missing check that the owner has DEFAULT_ADMIN_ROLE

File: scripts/acceptance_test_full_setup.py [hunk 13]
```diff
@@ -13,36 +14,30 @@
...
+    DEFAULT_ADMIN_ROLE
 )
 
 GRANT_ROLE_EVENT = "0x2f8788117e7eff1d82e926ec794901d17c78024a50270940304540a733656f0d"
 REVOKE_ROLE_EVENT = "0xf6391f5c32d9c69d2a47ea670b442974b53935d1edc7fd64eb21e047a839171b"
 
 deploy_config = deployment.AllowedRecipientsFullSetupDeployConfig(
-    token="",
-    limit=0,
+    tokens=["0xD87Ba7A50B2E7E660f678A895E4B72E7CB4CCd9C", "0x56340274fB5a72af1A3C6609061c451De7961Bd4"],
```
Comment (Psirex) [inline scripts/acceptance_test_full_setup.py:31]: Remove prefilled values

File: scripts/acceptance_test_full_setup.py [hunk 14]
```diff
@@ -83,70 +81,75 @@ def main():
...
     remove_allowed_recipient = RemoveAllowedRecipient.at(
         remove_allowed_recipient_address
     )
 
-    assert top_up_allowed_recipients.token() == deploy_config.token
-    assert top_up_allowed_recipients.allowedRecipientsRegistry() == registry
+    assert tokens_registry.getAllowedTokens() == deploy_config.tokens
+    assert len(tokens_registry.getAllowedTokens()) == len(deploy_config.tokens)
+
+    assert top_up_allowed_recipients.allowedRecipientsRegistry() == recipients_registry
```
Comment (Psirex) [inline scripts/acceptance_test_full_setup.py:102]: Missing check for allowedRecipientsTokens

File: scripts/acceptance_test_full_setup.py [hunk 15]
```diff
@@ -83,70 +81,75 @@ def main():
...
     assert registryPeriodDuration == deploy_config.period
 
     assert (
-        registry.spendableBalance() == deploy_config.limit - deploy_config.spent_amount
+        recipients_registry.spendableBalance() == deploy_config.limit - deploy_config.spent_amount
     )
 
-    assert registry.hasRole(ADD_RECIPIENT_TO_ALLOWED_LIST_ROLE, contracts.aragon.agent)
-    assert registry.hasRole(
+    assert recipients_registry.hasRole(ADD_RECIPIENT_TO_ALLOWED_LIST_ROLE, contracts.aragon.agent)
```
Comment (Psirex) [inline scripts/acceptance_test_full_setup.py:121]: No checks for allowed_tokens_registry

File: scripts/acceptance_test_single_setup.py [hunk 16]
```diff
@@ -1,41 +1,42 @@
...
 
 deploy_config = deployment.AllowedRecipientsSingleRecipientSetupDeployConfig(
-    period=0,
+    period=1,
     spent_amount=0,
-    title="",
-    limit=0,
-    token="",
-    trusted_caller="",
+    title="Test",
```
Comment (Psirex) [inline scripts/acceptance_test_single_setup.py:31]: Same as for full setup test

Comment (Psirex) [reply scripts/acceptance_test_single_setup.py:31]: For below checks comments are the same as for full setup test

File: scripts/deploy_allowed_tokens_factory.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    log,
+    deployed_date_time
+)
+
+from brownie import (
+    AllowedRecipientsFactory,
+    AllowedRecipientsBuilder
+)
+
+def main():
```
Comment (Psirex) [inline scripts/deploy_allowed_tokens_factory.py:23]: This script deploys nothing

File: utils/lido.py [hunk 18]
```diff
@@ -37,6 +37,22 @@ def addresses(network=DEFAULT_NETWORK):
     )
 
 
+def external_contracts(network=DEFAULT_NETWORK):
+    if network == "mainnet" or network == "mainnet-fork":
+        return {
```
Comment (TheDZhon) [inline utils/lido.py:42]: "usdt"?

**Review by TheDZhon — DISMISSED** [review]
LGTM :green_apple: 

Left some nitpicks and rough thoughts only.

Given the following:
- additional security measurements implemented with [LIP-13](https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-13.md) and preserved with this deployment as well
- only the DAO Agent can change the token list

The risks are tolerable and good to go from my side.
